### PR TITLE
 adding avro schema

### PR DESCRIPTION
**Pull request by jerrypeng**

### CODE REVIEW

**Comment by jerrypeng** [comment]
@mgodave please review

File: pulsar-client/pom.xml [hunk 1]
```diff
@@ -91,6 +91,12 @@
       </exclusions>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.avro</groupId>
+      <artifactId>avro</artifactId>
+      <version>1.8.2</version>
```
Comment (merlimat) [inline pulsar-client/pom.xml]: Version should be defined in top-level pom

Comment (jerrypeng) [reply pulsar-client/pom.xml]: ok

**Comment by jerrypeng** [comment]
@merlimat if you try a to deserialize a object with not the expected schema, an exception will be thrown

**Comment by jerrypeng** [comment]
retest this please

**Comment by merlimat** [comment]
> @merlimat if you try a to deserialize a object with not the expected schema, an exception will be thrown

It needs to prevent from creating a producer if the topic has a different schema. Otherwise it's not a "schema" but just a "serde"

**Comment by sijie** [comment]
@mgodave please review it.

**Comment by sijie** [comment]
> At least for Avro it would be good to have schema validation since it explicitely supports that.

I don't think there is a path to do a validation for now, unless I am not wrong. @mgodave ?

I think we should do the validation in a separate PR, more I think we need to think validation along with how SQL use schema registry.

**Comment by merlimat** [comment]
> I don't think there is a path to do a validation for now, unless I am not wrong. @mgodave ? 

I think there should be some kind of validation already for JSON. Also we need to validate that the schema is there. eg: topic is set to Avro<T>, a producer cannot connect and send "bytes". 

Avro has a way to extract schema info. That is what should be stored in the registry. Otherwise the whole "registry" part is completely useless.

> I think we should do the validation in a separate PR, more I think we need to think validation along with how SQL use schema registry.

Yes, if we don't store the "type" info, we cannot getting that back and discover how the data looks like

**Comment by sijie** [comment]
> Yes, if we don't store the "type" info, we cannot getting that back and discover how the data looks like

I think this PR already store the type info, no? It does same work as json schema. or I am missing something here?

**Comment by merlimat** [comment]
> I think this PR already store the type info, no? It does same work as json schema. or I am missing something here?

Sure, though I'm not sure where the validation is being done. We should have few unit tests to verify the behavior. eg what happens when : 
 * The topic is json and trying to publish avro
 * The avro schema for producer is not compatibile with existing avro schema
 * The avro schema represents an "allowed" evolution of schema
 * ..

**Comment by sijie** [comment]
@mgodave can you comment on [this question](https://github.com/apache/incubator-pulsar/pull/1917#issuecomment-394908822)? and do we have the test coverage?

**Comment by sijie** [comment]
> Avro specific validation on the server side

I image there should be a validation method in the `Schema` instance. so when broker get the schema info and retrieve the schema instance, it can use that schema instance to do validation. 

the validation should be part of `Schema` interface or a class in the schema package, and called in broker. but I think we are missing that, that's why I was suggesting us doing this in a separate PR.

**Comment by mgodave** [comment]
Implement [`SchemaCompatibilityCheck`](https://github.com/apache/incubator-pulsar/blob/db89cd460c25492bb6812f31a6e28265c33b192d/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaCompatibilityCheck.java) and add it to the [config file](https://github.com/apache/incubator-pulsar/blob/cd28ebfb6dbd1400d72a461d0c7790bddd0d6de4/pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java) for AVRO

**Comment by mgodave** [comment]
by default everything is "compatible", adding the checker allows you to customize that behavior.

**Comment by jerrypeng** [comment]
@merlimat @mgodave I have added the avro schema check. Though from testing, I can't seem to get the schema check to be invoked.  Not sure if there is a bug or I am doing some wrong

**Comment by jerrypeng** [comment]
retest this please

**Comment by jerrypeng** [comment]
@merlimat I have shaded the avro dependences. Please check if I got them all

File: pulsar-client-shaded/pom.xml [hunk 2]
```diff
@@ -86,6 +86,14 @@
                   <include>org.apache.httpcomponents:httpclient</include>
                   <include>commons-logging:commons-logging</include>
                   <include>org.apache.httpcomponents:httpcore</include>
+                  <include>org.apache.avro:avro</include>
```
Comment (merlimat) [inline pulsar-client-shaded/pom.xml]: @jerrypeng there's also pulsar-client-kafka :)

Comment (jerrypeng) [reply pulsar-client-shaded/pom.xml]: yup forgot you said that one as well

**Pull request merged**